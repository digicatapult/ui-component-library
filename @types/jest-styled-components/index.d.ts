declare module 'jest-styled-components/serializer'
