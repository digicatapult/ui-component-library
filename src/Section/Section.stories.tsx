import React from 'react'
import { Story } from '@storybook/react'

import Section from './index.js'

export default {
  title: 'Components/Section',
  component: Section,
  argTypes: {
    text: {
      control: {
        type: 'text',
      },
    },
    content: {
      control: {
        type: 'text',
      },
    },
    headingLevel: {
      control: {
        type: 'select',
        options: [1, 2, 3, 4, 5, 6],
      },
    },
    background: {
      control: {
        type: 'text',
      },
    },
    width: {
      control: {
        type: 'text',
      },
    },
    height: {
      control: {
        type: 'text',
      },
    },
    padding: {
      control: 'text',
    },
    headingSize: {
      control: 'text',
    },
    gap: {
      control: 'text',
    },
  },
}

const DefaultStoryTemplate: Story = (args) => {
  return (
    <Section
      headingLevel={args.headingLevel}
      title={args.text}
      width={args.width}
      height={args.height}
      padding={args.padding}
      headingGap={args.gap}
      headingSize={args.headingSize}
      background={args.background}
    >
      {args.content}
    </Section>
  )
}
export const Default = DefaultStoryTemplate.bind({})
Default.args = {
  headingLevel: 1,
  text: 'Hello',
  content: 'Some content',
}
export const HII = DefaultStoryTemplate.bind({})
HII.args = {
  headingLevel: 3,
  text: 'Project Details',
  content:
    'Scope / goal - The Â£33 million Low Carbon Hydrogen Supply competition aimed to accelerate the development of low carbon bulk hydrogen supply solutions in specific sectors. It was aimed at projects at a TRL of 4 to 7, which could result in lower capital or operating costs when compared to Steam Methane Reformation with Carbon Capture & Storage (SMR+CCS), or improve the carbon capture rates at a comparable cost.',
  width: '47ch',
  height: 'auto',
  padding: '1ch 4ch',
  headingSize: '1.2em',
  gap: '2ch',
  background: '#DFE667',
}
