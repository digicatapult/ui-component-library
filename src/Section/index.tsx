import React, {
  DetailedHTMLProps,
  HTMLAttributes,
  PropsWithChildren,
} from 'react'
import { useId } from 'react-id-generator'
import styled from 'styled-components'

import { HeadingLevel, HX } from '../Heading'

export interface SectionProps {
  headingLevel: HeadingLevel
  headingSize?: 'string'
  headingGap?: 'string'

  title: string

  background?: string
  padding?: string
  width?: string
  height?: string
}

interface SectionWrapperProps
  extends DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement> {
  background: string
  padding: string
  width: string
  height: string
}

interface SectionHeaderProps {
  size: string
  gap: string
}

const Section: React.FC<PropsWithChildren<SectionProps>> = ({
  title,
  children,
  headingLevel,
  headingGap = '0.5em',
  headingSize = '1.2em',
  background = '#f0f0f0',
  padding = '1em',
  width = '100ch',
  height = 'auto',
}) => {
  const [labelId] = useId()

  return (
    <Wrapper
      aria-labelledby={labelId}
      background={background}
      padding={padding}
      width={width}
      height={height}
    >
      <Heading headingLevel={headingLevel} size={headingSize} gap={headingGap}>
        {title}
      </Heading>
      {children}
    </Wrapper>
  )
}

const Wrapper = styled.section<SectionWrapperProps>`
  background: ${({ background }) => background};
  padding: ${({ padding }) => padding};
  width: ${({ width }) => width};
  height: ${({ height }) => height};
`

const Heading = styled(HX)<SectionHeaderProps>`
  font-size: ${({ size }) => size};
  margin: 0;
  padding: 0;
  margin-bottom: ${({ gap }) => gap};
`

export default Section
