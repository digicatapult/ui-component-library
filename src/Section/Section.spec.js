import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Section from './index.tsx'

describe('Section', () => {
  test('default', () => {
    const tree = renderer
      .create(
        <Section headingLevel={1} title="TITLE">
          Body content goes here
        </Section>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('overrides', () => {
    const tree = renderer
      .create(
        <Section
          headingLevel={1}
          title="TITLE"
          width="20ch"
          height="100%"
          padding="5ch"
          headingGap="1ch"
          background="blue"
        >
          Body content goes here
        </Section>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
