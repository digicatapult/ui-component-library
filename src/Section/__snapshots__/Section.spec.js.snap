// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section default 1`] = `
.c1 {
  font-size: 1em;
  padding: 0;
  margin: 0;
}

.c0 {
  background: #f0f0f0;
  padding: 1em;
  width: 100ch;
  height: auto;
}

.c2 {
  font-size: 1.2em;
  margin: 0;
  padding: 0;
  margin-bottom: 0.5em;
}

<section
  aria-labelledby="id1"
  className="c0"
  height="auto"
  width="100ch"
>
  <h1
    className="c1 c2"
    size="1.2em"
  >
    TITLE
  </h1>
  Body content goes here
</section>
`;

exports[`Section overrides 1`] = `
.c1 {
  font-size: 1em;
  padding: 0;
  margin: 0;
}

.c0 {
  background: blue;
  padding: 5ch;
  width: 20ch;
  height: 100%;
}

.c2 {
  font-size: 1.2em;
  margin: 0;
  padding: 0;
  margin-bottom: 1ch;
}

<section
  aria-labelledby="id2"
  className="c0"
  height="100%"
  width="20ch"
>
  <h1
    className="c1 c2"
    size="1.2em"
  >
    TITLE
  </h1>
  Body content goes here
</section>
`;
